### ISSUE TITLE
How to replace the loss function？

### ISSUE THREAD

**Issue by quuhua911** [awslabs/autogluon#241]
I would like to replace softmax with A-softmax or L-softmax, where I need to modify the parameters？
My problem_type is "multiclass" ,data in tabular format .

**Comment by jwmueller** [comment]
This may be challenging to do in an appropriate manner, and I wouldn't recommend it unless you have a very strong reason to do so.

For training, you probably are hoping to change the loss function for the TabularNeuralNet model 
 (I'm not sure you can use this loss with all the other models that AutoGluon also automatically fits to tabular data).  Because AutoGluon automatically ensembles multiple other models with the neural network, you may not notice a drastic difference only changing the neural network training objective, unless you tell `fit()` that it should only train a neural network model by specifying `hyperparams = {'NN': {...}}` as a dict where 'NN' is the only key.

To change the neural net training objective, you can replace where the softmax loss is used here:  https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py#L513

I'm not sure if your desired losses are available in MXNet, otherwise you'll need to define the loss functions yourself, following the MXNet guidelines for custom loss functions.  

You can also probably specify your custom loss to the lightGBM model trained by AutoGluon in this file:
https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/lgb/lgb_model.py#L91
by following these instructions: 
https://towardsdatascience.com/custom-loss-functions-for-gradient-boosting-f79c1b40466d

And you may be able to repeat this for the CatBoost model trained by AutoGluon in this file:
https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/catboost/catboost_model.py
by following these instructions:  https://github.com/catboost/catboost/issues/112

Finally, you may also want to specify `eval_metric` to match your custom-loss (used for ensemble-selection, early-stopping, HPO, etc).  In this case, you will need to define a new metric function, following the format of the metrics defined in:  https://github.com/awslabs/autogluon/tree/master/autogluon/utils/tabular/metrics
